{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# CyC-production COVID-19 UKB analyses"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extract phenotype\n",
    "\n",
    "COVID-19 phenotype data downloaded from UKB data portal on June 23rd 2021"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(reshape2)\n",
    "library(dplyr)\n",
    "library(tidyr)\n",
    "library(tidyverse)\n",
    "library(stringr)\n",
    "library(survival)\n",
    "library(fst)\n",
    "\n",
    "#COVID tests\n",
    "library(readr)\n",
    "covid19_result_england <- read_delim(\"covid19_result_england.txt\", \n",
    "                                     \"\\t\", escape_double = FALSE, trim_ws = TRUE)\n",
    "pos_england = subset(covid19_result_england, result==1)$eid\n",
    "\n",
    "covid19_result_scotland <- read_delim(\"covid19_result_scotland.txt\", \n",
    "                                      \"\\t\", escape_double = FALSE, trim_ws = TRUE)\n",
    "pos_scotland = subset(covid19_result_scotland, result==1)$eid\n",
    "\n",
    "covid19_result_wales <- read_delim(\"covid19_result_wales.txt\", \n",
    "                                   \"\\t\", escape_double = FALSE, trim_ws = TRUE)\n",
    "\n",
    "pos_wales = subset(covid19_result_wales, result==1)$eid\n",
    "pos_all = unique(c(pos_england, pos_scotland, pos_wales))\n",
    "\n",
    "#Hospital-coded\n",
    "hesin_diag <- read_delim(\"hesin_diag.txt\", \n",
    "                         \"\\t\", escape_double = FALSE, trim_ws = TRUE)\n",
    "metabolic<-c(\"U071\")\n",
    "hesin_diag = subset(hesin_diag, diag_icd10 %in% metabolic & level==1)\n",
    "hosp_all = unique(hesin_diag$eid)\n",
    "\n",
    "#In-hospital mortality\n",
    "hesin <- read_delim(\"hesin.txt\", \n",
    "                         \"\\t\", escape_double = FALSE, trim_ws = TRUE)\n",
    "hesin = inner_join(hesin, subset(hesin_diag, level==1), by=c('eid','ins_index'))\n",
    "hesin = subset(hesin, disdest_uni==\"11001\")\n",
    "\n",
    "death_in_hospital = unique(hesin$eid)\n",
    "\n",
    "\n",
    "#Critical care admission\n",
    "hesin_critical <- read_delim(\"hesin_critical.txt\", \n",
    "                             \"\\t\", escape_double = FALSE, trim_ws = TRUE)\n",
    "\n",
    "library(dplyr)\n",
    "hesin_critical = inner_join(hesin_critical, subset(hesin_diag, level==1), by=c('eid','ins_index'))\n",
    "critical_all = unique(hesin_critical$eid)\n",
    "\n",
    "#Deaths\n",
    "death_cause <- read_delim(\"death_cause.txt\", \n",
    "                          \"\\t\", escape_double = FALSE, trim_ws = TRUE)\n",
    "death_cause = subset(death_cause, cause_icd10 %in% metabolic & level==1)\n",
    "death_all = unique(death_cause$eid)\n",
    "death_all = unique(c(subset(hesin_critical, ccdisdest==6)$eid, death_all, death_in_hospital))\n",
    "\n",
    "death_and_hospital = intersect(death_all, hosp_all)\n",
    "\n",
    "#Total cases\n",
    "total = unique(c(pos_all, hosp_all, critical_all, death_all))\n",
    "covid = data.frame(eid=total)\n",
    "covid$hospital = ifelse(covid$eid %in% hosp_all,1,0)\n",
    "covid$critical = ifelse(covid$eid %in% critical_all,1,0)\n",
    "covid$death = ifelse(covid$eid %in% death_all,1,0)\n",
    "covid$death_hospital = ifelse(covid$eid %in% c(death_and_hospital,death_in_hospital),1,0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Regression - COVID-19 outcomes against CyC-production PGS\n",
    "\n",
    "CyC-production PGS was scored as described previously using PLINK2, the polygenic score SNP coefficients will be uploaded to PGS Catalog. Covariates included age, sex, genotyping array, recruitment centre and principal components 1-10 which were computed on an list of LD-pruned SNPs provided by UKB ('ukb_snp_qc.txt')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Hail/Python script for computing principal components in EUR ancestry cohort"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import hail as hl\n",
    "import os\n",
    "import pandas as pd\n",
    "import subprocess\n",
    "from gnomad.sample_qc.ancestry import *\n",
    "\n",
    "#Define memory and CPU availability\n",
    "\n",
    "tmp = \"/mnt/grid/janowitz/rdata_norepl/tmp\"\n",
    "\n",
    "os.environ[\"SPARK_LOCAL_DIRS\"]=tmp\n",
    "os.environ[\"PYSPARK_SUBMIT_ARGS\"] =\"--conf spark.network.timeout=15m --conf spark.executor.heartbeatInterval=10m --conf spark.memory.fraction=1.0 --driver-memory 2880G pyspark-shell\"\n",
    "\n",
    "\n",
    "hl.init(default_reference='GRCh37', master ='local[96]',min_block_size=128, local_tmpdir=tmp, tmp_dir=tmp)\n",
    "\n",
    "\n",
    "#Pre-process ancestry predictions, merge with phenotype data\n",
    "ukb_predictions = hl.read_table(\"ukb_predictions.ht\")\n",
    "\n",
    "ukb_predictions_pd = ukb_predictions.to_pandas()\n",
    "ukb_predictions_pd = ukb_predictions_pd[[\"s\", \"pop\"]]\n",
    "ukb_predictions_pd = ukb_predictions_pd[ukb_predictions_pd['pop'] != \"Other\"]\n",
    "print(ukb_predictions_pd['pop'].value_counts(), flush=True)\n",
    "\n",
    "phenotypes = pd.read_csv('phenotypes.csv')\n",
    "phenotypes = phenotypes.drop(phenotypes.columns[0], axis=1)\n",
    "phenotypes = phenotypes.rename(columns={'id': 's'})\n",
    "phenotypes['s'] = phenotypes.s.astype(int)\n",
    "ukb_predictions_pd['s'] = ukb_predictions_pd.s.astype(int)\n",
    "\n",
    "phenotypes = phenotypes.merge(ukb_predictions_pd, how='inner', on='s')\n",
    "\n",
    "populations = phenotypes['pop'].unique().tolist()\n",
    "#populations = [\"CSA\"]\n",
    "\n",
    "#Add relationship matrix (UKB-provided KING analysis)\n",
    "rel = hl.import_table('ukbXXXXX_rel_sXXXXX.dat', impute=True,delimiter='\\s+', \n",
    "                     types={'ID1': hl.tstr, 'ID2': hl.tstr})\n",
    "rel = rel.select(rel.ID1, rel.ID2, rel.Kinship)\n",
    "\n",
    "#Import GRCh37 data (filtered to call rate, MAF, no H-W filtering at present)\n",
    "ukb_gwas = hl.read_matrix_table('ukb_grch37_filtered_forgwas.mt')\n",
    "ukb_gwas = ukb_gwas.filter_rows(ukb_gwas.variant_qc.AC[1] > 100) #As per Regenie documentation\n",
    "print(ukb_gwas.count(), flush=True)\n",
    "\n",
    "#Prune GRCh37 data using SNP list (in_PCA) from marker paper\n",
    "snp_ukb = hl.import_table('ukb_snp_qc.txt', impute=True, delimiter='\\s+')\n",
    "snp_ukb = snp_ukb.filter(snp_ukb.in_PCA==1)\n",
    "snp_ukb = snp_ukb.key_by(snp_ukb.rs_id)\n",
    "ukb_gwas_pruned = ukb_gwas.filter_rows(hl.is_defined(snp_ukb[ukb_gwas.rsid]))\n",
    "\n",
    "for pop in populations:\n",
    "    print(pop, flush=True)\n",
    "    phenotypes['s'] = phenotypes.s.astype(str)\n",
    "    \n",
    "    #Subset covariate file, make suitable for BOLT-LMM\n",
    "    \n",
    "    subset = phenotypes[phenotypes['pop'] == pop]\n",
    "    \n",
    "    frame = { 'FID': subset['s'], 'IID': subset['s'], 'year_of_birth': subset['year_of_birth_f34_0_0'],\n",
    "             'sex': subset['sex_f31_0_0'],'array': subset['genotype_measurement_batch_f22000_0_0'],\n",
    "        'centre': subset['uk_biobank_assessment_centre_f54_0_0']}\n",
    "    \n",
    "    #Subset GRCH37 version of UK biobank dataset\n",
    "    \n",
    "    subset_ht = hl.Table.from_pandas(subset['s'].to_frame(), key='s') \n",
    "    ukb_pop_gwas = ukb_gwas.filter_cols(hl.is_defined(subset_ht[ukb_gwas.col_key]))\n",
    "    ukb_pop_gwas = hl.variant_qc(ukb_pop_gwas)\n",
    "    ukb_pop_gwas = ukb_pop_gwas.filter_rows(ukb_pop_gwas.variant_qc.AC[1] > 20) #As per Regenie documentation\n",
    "    ukb_pop_gwas_pruned = ukb_gwas_pruned.filter_cols(hl.is_defined(subset_ht[ukb_gwas_pruned.col_key]))\n",
    "\n",
    "    #Run PCA on all filtered variants, per population, removing relateds subjects per population\n",
    "    \n",
    "    rel_sub = rel.filter(hl.is_defined(subset_ht[rel.ID1]) & hl.is_defined(subset_ht[rel.ID2]))\n",
    "    pairs = rel_sub.filter(rel_sub['Kinship'] > 0.0442) #Third degree relatives per KING documentation\n",
    "    related_samples_to_remove = hl.maximal_independent_set(pairs.ID1, pairs.ID2, False)\n",
    "    \n",
    "    print(\"Related + \", pop, related_samples_to_remove.count(), flush=True)\n",
    "    \n",
    "    _, scores_pca_ref, _ = run_pca_with_relateds(ukb_pop_gwas_pruned, related_samples_to_remove, \n",
    "                                                               n_pcs=10, autosomes_only=True)\n",
    "    \n",
    "    scores_pca_ref = scores_pca_ref.transmute(**{f'PC{i}': scores_pca_ref.scores[i - 1] for i in range(1, 11)})\n",
    "    scores_pd = scores_pca_ref.to_pandas()\n",
    "    scores_pd = scores_pd.rename(columns={'s': 'IID'})\n",
    "    \n",
    "    #Add PCA results to covariates file, then save\n",
    "    \n",
    "    covariates = covariates.merge(scores_pd, left_on='IID', right_on='IID')\n",
    "    folder = 'PATH' + pop + '/covariates.tsv'\n",
    "    covariates.to_csv(folder, sep=\"\\t\", index=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Regression (R script)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(readr)\n",
    "library(dplyr)\n",
    "\n",
    "scale2 <- function(x, na.rm = TRUE) (x - mean(x, na.rm = na.rm)) / sd(x, na.rm)\n",
    "PGS_1e4_RAPIDO <- read_table2(\"UKB_UKB380_PGS_inner.sscore\")\n",
    "\n",
    "\n",
    "\n",
    "names(PGS_1e4_RAPIDO)[2]<-\"id\"\n",
    "PGS_1e4_RAPIDO$SCORE = PGS_1e4_RAPIDO$SCORE1_AVG\n",
    "\n",
    "#Import additional variables\n",
    "populations <- read_csv(\"populations.csv\")\n",
    "PGS_1e4_RAPIDO = subset(PGS_1e4_RAPIDO, id %in% subset(populations, pop==\"EUR\")$s)\n",
    "\n",
    "#Filter to cohort of patients ('validation cohort') who were not used for construct of the polygenic score\n",
    "validation_cohort <- read_delim(\"validation_cohort.tsv\", \n",
    "                                \"\\t\", escape_double = FALSE, trim_ws = TRUE)\n",
    "\n",
    "covariates <- read_table2(\"covariates_eur.tsv\")\n",
    "covariates = covariates[,-4]\n",
    "names(covariates)[2]<-\"id\"\n",
    "\n",
    "\n",
    "PGS_1e4_RAPIDO = subset(PGS_1e4_RAPIDO, id %in% validation_cohort$s)\n",
    "PGS_1e4_RAPIDO = inner_join(PGS_1e4_RAPIDO, covariates, by=\"id\")\n",
    "\n",
    "#Remove any duplicated cystatin measurements, pick earliest measurement\n",
    "\n",
    "#Normalize score\n",
    "PGS_1e4_RAPIDO$score_scale = scale2(PGS_1e4_RAPIDO$SCORE)\n",
    "PGS_1e4_RAPIDO$score_decile <- as.factor(ntile(PGS_1e4_RAPIDO$SCORE, 10))\n",
    "PGS_1e4_RAPIDO$score_quartile <- as.factor(ntile(PGS_1e4_RAPIDO$SCORE, 4))\n",
    "\n",
    "\n",
    "names(PGS_1e4_RAPIDO)[2]='eid'\n",
    "covidx <- inner_join(covid, PGS_1e4_RAPIDO, by='eid')\n",
    "m<-match(covidx$eid, survival$id)\n",
    "covidx$sex = ifelse(survival$sex[m]==\"Male\",1,0)\n",
    "covidx$bmi = survival$bmi[m]\n",
    "\n",
    "covidx$control = ifelse(covidx$hospital==0 & covidx$critical==0 & covidx$death==0,1,0)\n",
    "\n",
    "\n",
    "comb = data.frame()\n",
    "mylogit = glm(formula = hospital ~ score_scale + sex + PC1 + PC2 + PC3 + PC4 + PC5 + PC6 + PC7 + PC8 + PC9 + PC10 + year_of_birth + centre + array, family = \"binomial\", data=subset(covidx, hospital ==1 | control==1))\n",
    "add = data.frame(comparison=\"hospitalization\",  HR = exp(mylogit$coefficients[2]), lower = exp(confint(mylogit, level=0.95)[2,1]),\n",
    "                 upper = exp(confint(mylogit, level=0.95)[2,2]), p_value=summary(mylogit)$coefficients[2,4])\n",
    "comb = rbind(comb, add)\n",
    "\n",
    "mylogit = glm(formula = critical ~ score_scale + sex + PC1 + PC2 + PC3 + PC4 + PC5 + PC6 + PC7 + PC8 + PC9 + PC10 + year_of_birth + centre + array, family = \"binomial\", data=subset(covidx, critical ==1 | control==1))\n",
    "add = data.frame(comparison=\"critical care\",  HR = exp(mylogit$coefficients[2]), lower = exp(confint(mylogit, level=0.95)[2,1]),\n",
    "                 upper = exp(confint(mylogit, level=0.95)[2,2]), p_value=summary(mylogit)$coefficients[2,4])\n",
    "comb = rbind(comb, add)\n",
    "\n",
    "mylogit = glm(formula = death ~ score_scale + sex + PC1 + PC2 + PC3 + PC4 + PC5 + PC6 + PC7 + PC8 + PC9 + PC10 + year_of_birth + centre + array, family = \"binomial\",data=subset(covidx, death ==1 | control==1))\n",
    "add = data.frame(comparison=\"death\",  HR = exp(mylogit$coefficients[2]), lower = exp(confint(mylogit, level=0.95)[2,1]),\n",
    "                 upper = exp(confint(mylogit, level=0.95)[2,2]), p_value=summary(mylogit)$coefficients[2,4])\n",
    "comb = rbind(comb, add)\n",
    "\n",
    "comb %>% dplyr::mutate(comparison = factor(comparison, levels = c(\"death\", \"critical care\",\"hospitalization\"))) %>% ggplot(aes(x = comparison,y = HR, ymin = lower, ymax = upper ))+\n",
    "  geom_pointrange(aes(col=comparison))+\n",
    "  geom_hline(aes(fill=comparison),yintercept =1, linetype=2)+\n",
    "  xlab('Group')+ ylab(\"Odds Ratio (95% Confidence Interval)\")+\n",
    "  geom_errorbar(aes(ymin=lower, ymax=upper,col=comparison),width=0.5,cex=1)+ \n",
    "  theme(plot.title=element_text(size=16,face=\"bold\"),\n",
    "        axis.text.y=element_blank(),\n",
    "        axis.ticks.y=element_blank(),\n",
    "        axis.text.x=element_text(face=\"bold\"),\n",
    "        axis.title=element_text(size=12,face=\"bold\"),\n",
    "        strip.text.y = element_text(hjust=0,vjust = 1,angle=180,face=\"bold\"))+\n",
    "  coord_flip()+theme_bw()+xlab(\"\")+ theme(legend.position = \"none\") "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Population controls analysis (R script)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "use = PGS_1e4_RAPIDO\n",
    "m<-match(use$eid, survival$id)\n",
    "use$sex = ifelse(survival$sex[m]==\"Male\",1,0)\n",
    "use$bmi = survival$bmi[m]\n",
    "\n",
    "\n",
    "controls = subset(use, !(eid %in% covid$eid))\n",
    "controls$stat = 0\n",
    "cases = subset(use, eid %in% subset(covid, critical==1)$eid)\n",
    "cases$stat = 1\n",
    "\n",
    "analysis = rbind(controls, cases)\n",
    "mylogit = glm(formula = stat ~ score_scale + bmi + sex + PC1 + PC2 + PC3 + PC4 + PC5 + PC6 + PC7 + PC8 + PC9 + PC10 + year_of_birth + centre + array, family = \"binomial\", data=analysis)\n",
    "summary(mylogit)\n",
    "add = data.frame(comparison=\"hospitalization\",  HR = exp(mylogit$coefficients[2]), lower = exp(confint(mylogit, level=0.95)[2,1]),\n",
    "                 upper = exp(confint(mylogit, level=0.95)[2,2]))\n",
    "print(add)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Charite Hospital C2 ratio analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(dplyr)\n",
    "library(ggplot2)\n",
    "library(readr)\n",
    "library(readxl)\n",
    "data <- read_delim(\"data.txt\", \n",
    "                   \"\\t\", escape_double = FALSE, trim_ws = TRUE)\n",
    "\n",
    "metadata <- read_delim(\"metadata.txt\", \n",
    "                       \"\\t\", escape_double = FALSE, trim_ws = TRUE)\n",
    "\n",
    "new_data <- read_excel(\"20220209_data_long_upd.xlsx\", sheet=\"minimal\")\n",
    "new_data$patient2 = paste('00',new_data$patient,sep='')\n",
    "\n",
    "\n",
    "process = as.data.frame(t(data[,-1]))\n",
    "names(process) = data$Features\n",
    "\n",
    "process$patient<-row.names(process)\n",
    "\n",
    "names(metadata)[1]<-\"patient\"\n",
    "process = inner_join(metadata, process, by=\"patient\")\n",
    "process=process[,c(1,2:19,49,50, 96, 51, 97, 84, 98, 99,53,36)]\n",
    "\n",
    "m<-match(process$patient,new_data$patient)\n",
    "m2<-match(process$patient,new_data$patient2)\n",
    "m3=coalesce(m, m2)\n",
    "\n",
    "process$creatinine_new=new_data$`Creatinine, mg/dl`[m3]\n",
    "process$crp_new=new_data$`CRP (clinical), mg/l`[m3]\n",
    "process$bmi_new=new_data$BMI[m3]\n",
    "\n",
    "\n",
    "\n",
    "process$`Creatinine, mg/dl`=coalesce(process$`Creatinine, mg/dl`, process$creatinine_new)\n",
    "\n",
    "process = process %>% arrange(SP.Date) %>% group_by(Id) %>% tidyr::fill(`Creatinine, mg/dl`, .direction=\"downup\") #Impute with most recent value\n",
    "process = process %>% arrange(SP.Date) %>% group_by(Id) %>% tidyr::fill(`CRP (clinical), mg/l`, .direction=\"downup\") #Impute with most recent value\n",
    "process = process %>% arrange(SP.Date) %>% group_by(Id) %>% tidyr::fill(CST3, .direction=\"downup\") #Impute with most recent value\n",
    "\n",
    "\n",
    "process$`Creatinine, mg/dl`[is.nan(process$`Creatinine, mg/dl`)] <- NA\n",
    "\n",
    "process = subset(process, is.na(`Creatinine, mg/dl`)==FALSE)\n",
    "\n",
    "process$outcome_date = process$SP.Date - process$I.Days.Delta.Outcome\n",
    "\n",
    "process$os_days = NA\n",
    "\n",
    "for(i in 1:nrow(process)) {\n",
    "  \n",
    "  row = process[i,]\n",
    "  if(row$I.Died ==1) {process$os_days[i] = row$outcome_date - row$SP.Date}\n",
    "  if(row$I.Died ==0) {process$os_days[i] = as.Date(\"2020-12-30\") - row$SP.Date}\n",
    "  \n",
    "}\n",
    "\n",
    "process$os_stat = process$I.Died\n",
    "\n",
    "library(nephro)\n",
    "library(tidyr)\n",
    "\n",
    "process$ethnicity<-0\n",
    "\n",
    "process = process %>% tidyr::drop_na(c(\"F.Sex\",\"Creatinine, mg/dl\", \"I.Age\",\"CST3\", \"BMI\")) %>% mutate(F.Sex = ifelse(F.Sex==\"MaNNLICH\",1,0)) %>% mutate(egfr_creatinine=CKDEpi.creat(`Creatinine, mg/dl`,F.Sex, I.Age, ethnicity),egfr_cystatin = CKDEpi.cys((CST3/300),F.Sex, I.Age))\n",
    "\n",
    "model=lm(egfr_cystatin ~ egfr_creatinine, data=process)\n",
    "process$cortiscore = -resid(model)\n",
    "\n",
    "process$ratio = process$egfr_creatinine / process$egfr_cystatin\n",
    "ggplot(process, aes(x=as.factor(I.After.Dexa), y=cortiscore))+geom_jitter(width=0.1)\n",
    "\n",
    "\n",
    "library(\"survival\")\n",
    "library(\"survminer\")\n",
    "library(finalfit)\n",
    "\n",
    "process$cortistat = ifelse(process$ratio >1, 1, 0)\n",
    "\n",
    "#process = process %>% group_by(Id) %>% slice_min(order_by=SP.Date, n=1)\n",
    "\n",
    "process = process  %>% mutate(F.Sex = as.factor(F.Sex), CRP = `CRP (clinical), mg/l`, charlson = `Charlson score`)\n",
    "\n",
    "\n",
    "\n",
    "library(coxme)\n",
    "\n",
    "m2 <- coxph(Surv(os_days, os_stat) ~ I.Age + F.Sex + BMI + cortistat + CRP + charlson + cluster(Id), \n",
    "            data = process)\n",
    "\n",
    "\n",
    "m3 <- coxph(Surv(os_days, os_stat) ~ cortistat + cluster(Id), \n",
    "            data = process)\n",
    "\n",
    "\n",
    "summary(m2)\n",
    "summary(m3)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R/4.0.3",
   "language": "R",
   "name": "r_4.0.3"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
